### PR TITLE
Telemetry

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  private class TelemetryJob extends Job {
+
+    public TelemetryJob() {
+      super("SonarLint Telemetry");
+    }
+
+    protected IStatus run(IProgressMonitor monitor) {
+      schedule(6 * 60 * 60 * 1000L);
```
Comment (dbmeneses) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java]: Might be clearer TimeUnit.HOURS.toMillis(6)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      super("SonarLint Telemetry");
+    }
+
+    protected IStatus run(IProgressMonitor monitor) {
+      schedule(TimeUnit.HOURS.toMillis(6));
+      if (enabled) {
+        try {
+          TelemetryClientConfig clientConfig = getTelemetryClientConfig();
+          telemetryEngine.getClient().tryUpload(clientConfig, isAnyProjectConnected());
+        } catch (Exception e) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    if (telemetryEngine != null) {
+      if (optOut == !telemetryEngine.enabled()) {
+        return;
+      }
+      telemetryEngine.enable(!optOut);
+      if (optOut) {
+        try {
+          TelemetryClientConfig clientConfig = getTelemetryClientConfig();
+          telemetryEngine.getClient().optOut(clientConfig, isAnyProjectConnected());
+        } catch (Exception e) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    SonarLintCorePlugin.getAnalysisListenerManager().removeListener(this);
+    if (scheduledJob != null) {
+      scheduledJob.cancel();
+      scheduledJob = null;
+    }
+    try {
+      if (telemetryEngine != null) {
+        telemetryEngine.save();
+      }
+    } catch (IOException e) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+      SonarLintLogger.get().info("Telemetry disabled by system property");
+      return;
+    }
+    try {
+      this.telemetryEngine = new Telemetry(getStorageFilePath(), PRODUCT, SonarLintUtils.getPluginVersion());
+      SonarLintCorePlugin.getAnalysisListenerManager().addListener(this);
+      this.scheduledJob = new TelemetryJob();
+      scheduledJob.schedule(TimeUnit.MINUTES.toMillis(1));
+      this.enabled = true;
+    } catch (Exception e) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java]: MAJOR Either log or rethrow this exception. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**